### PR TITLE
CI: unify environment creation 

### LINKED ISSUE #23923: TST: Review if we are really testing with different locales
In the scripts for the CI, we set the locale with:
```
export LC_ALL="$LOCALE_OVERRIDE";
export LANG="$LOCALE_OVERRIDE";
```
and immediately later we display the locale detected by pandas:
```
pycmd='import pandas; print("pandas detected console encoding: %s" % pandas.get_option("display.encoding"))'
python -c "$pycmd"
```
(see https://github.com/pandas-dev/pandas/blob/master/ci/script_single.sh#L5)

Checking the logs, I can see how in the two builds where we specify locales, the output is:
```
# Linux py37_locale
2018-11-25T20:35:30.0051096Z Setting LC_ALL and LANG to zh_CN.UTF-8
2018-11-25T20:35:31.0423116Z pandas detected console encoding: UTF-8
```
and
```
# Linux py36_locale_slow
2018-11-25T20:37:45.3938178Z Setting LC_ALL and LANG to it_IT.UTF-8
2018-11-25T20:37:46.4791772Z pandas detected console encoding: UTF-8
```

And also in Travis:
```
# 2.7, locale, slow, old NumPy
Setting LC_ALL and LANG to zh_CN.UTF-8
pandas detected console encoding: UTF-8
```

I checked the logs for the PR https://github.com/pandas-dev/pandas/pull/23908 but I assume all are the same.

If I'm not misunderstanding anything, I think pandas is not detecting the locale we want to test with, and it's testing with the default `UTF8` locale, as in the rest of the tests.

CC: @jreback @TomAugspurger

### LINKED ISSUE #24498: CI: Simplify and unify environment creation
When creating a pandas environment for the CI, we currently have two separate systems, one for Travis and one for Azure. Summarizing, those are the involved steps.

**travis**
```
ci/prep_cython_cache.sh
ci/install_travis.sh         -> locale, download/install conda, ccache, create env, setup.py develop
ci/submit_cython_cache.sh
ci/install_db_travis.sh      -> create dbs
ci/before_script_travis.sh   -> xvfb (clipboard)
```

**azure**
```
ci/incremental/install_miniconda.sh          -> download/install conda
ci/incremental/setup_conda_environment.sh    -> create env, locale
ci/incremental/build.sh                      -> setup.py build_ext
```

Except for the cython cache, I think all the rest could be unified in a single file `ci/setup_env.sh` that performs all the steps.

This would simplify and avoid duplicate code and inconsistencies. As well as making future changes simpler.

### PR DESCRIPTION
**Pull request by saurav2608**
closes #24498
closes #23923

### CODE REVIEW

**Comment by saurav2608** [comment]
This sets up DB in both travis and azure. Have not tested DB jobs on Azure but maybe worth a try.

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: this is not needed (and put the set -v -e to the top)

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: remove this

**Review by jreback — CHANGES_REQUESTED** [review]
needs some edits

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: remove the -v, put the -e on the bash line

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: @datapythonista how is this supposed to work on windows?

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: put a blank line here (and generally between sections)

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: I *think* this can be in edit_init

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: superfluous

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: extra

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: move the done to the appropratea places below

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
yes remove the original files

**Review by datapythonista — COMMENTED** [review]
besides the comments, there is an inconsistent use of the `echo` commands, with cases like:
- `echo Error`
- `echo "done"`
- `echo "[done]"`

And some times with a bare `echo` before and some without (I guess those add blank lines).

I'd be consistent and use the same format everywhere. I don't think we are using the squared brackets in any other script. So, I'd simply use: `echo "some message"` everywhere. But feel free to check the log and pick the format that makes more sense, but use the same everywhere.

File: ci/setup_env.sh
Comment (saurav2608) [reply ci/setup_env.sh]: No this is not replacing `ci/prep_cython_cache.sh`. This part of code was present in `ci/install_travis.sh`.

**Review by datapythonista — CHANGES_REQUESTED** [review]
Check azure, your script is failing.

**Comment by saurav2608** [comment]
@datapythonista  - Azure is failing on conda update specifically - `conda update -q conda`. I was reading [here](https://github.com/conda/conda/issues/3869#issuecomment-384781590) that conda now suggests `conda update -n base conda`. I can change the script to reflect this. However, why are we even doing this? We are downloading and installing the latest miniconda build. Why would we need to update conda just after that. 

Another question is on line 32 of `setup_env.sh`. Why is the `rm -rf ` is needed in the current setup. I would think that the VM running the tests is spun fresh everytime we start the test.

**Comment by datapythonista** [comment]
Yes, please update the conda update statement to the new syntax. Anaconda can release a new conda version, but not create a new installable immediately, that's why it's common to update conda itself after installation. It's similar to when you install Linux and immediately run `dnf update` or equivalent. The iso is older than the versions of the latest packages.

I'm not sure about the `rm -rf`. I guess travis can reuse the containers or something. If the environment is clean for every build, it surely doesn't make sense. But please leave it for now, I'm sure it was added for a reason.

File: ci/setup_env.sh
Comment (saurav2608) [reply ci/setup_env.sh]: Setting DB for Azure did not work. I think the base environment should have mySQL installed. I changed the if to include only travis builds.

**Comment by saurav2608** [comment]
Pending items -

- DB works only on Travis
- Not sure where which test required xvfb. But I could not start it on Azure. I do not think this was enabled earlier as well.

File: ci/setup_env.sh
Comment (datapythonista) [inline ci/setup_env.sh]: I don't see the value of having this in a function, as it's called just once and immediately after defining it. Can you simply run this code directly?

File: ci/setup_env.sh
Comment (datapythonista) [inline ci/setup_env.sh]: I think we removed the squared brackets from everywhere except this block. Can you remove them from here too please?

File: ci/setup_env.sh
Comment (datapythonista) [inline ci/setup_env.sh]: make `sedc` uppercase, as the rest of the variables.

File: ci/setup_env.sh
Comment (datapythonista) [inline ci/setup_env.sh]: `HOME_DIR` is not used, I would remove these 3 lines, I don't think the echo is really useful.

File: ci/setup_env.sh
Comment (datapythonista) [inline ci/setup_env.sh]:
```suggestion
    echo "rm -rf $MINICONDA_DIR"
```

File: ci/setup_env.sh
Comment (datapythonista) [inline ci/setup_env.sh]: use `$MINICONDA_DIR` instead of building the path again.

File: azure-pipelines.yml
Comment (datapythonista) [inline azure-pipelines.yml]:
```suggestion
    displayName: 'Setup environment and build pandas'
```

For consistency with the other azure build

**Review by datapythonista — COMMENTED** [review]
looks good, just few minor things

**Review by datapythonista — COMMENTED** [review]
lgtm, thanks @saurav2608 

@jreback @TomAugspurger can you take a look?

**Comment by jreback** [comment]
I am not inclinded to merge this until *after* @TomAugspurger cuts the RC (haven't really looked yet). but the CI churn recently has caused too many issues.

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: so how does this work on windows?

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: @datapythonista I think nowadays this caching might be counter-productive, though needs some testing, if you want to make an issue to remove it (but don't touch here)

File: ci/setup_env.sh
Comment (jreback) [inline ci/setup_env.sh]: this also maybe can removed in the future (not here) @datapythonista

**Review by jreback — CHANGES_REQUESTED** [review]

File: ci/setup_env.sh
Comment (datapythonista) [reply ci/setup_env.sh]: Windows can't use this script unfortunately, as it's a shell script. In Windows, we have `ci/incremental/setup_conda_environment.cmd` for setting up the environment, and conda is downloaded directly in the azure configuration with powershell commands.

**Review by datapythonista — APPROVED** [review]

**Comment by saurav2608** [comment]
@datapythonista / @jreback - any thoughts on this PR?

**Comment by jreback** [comment]
can you merge master and i'll loo again here

**Comment by saurav2608** [comment]
@jreback - green.

**Comment by jreback** [comment]
@saurav2608 pls confirm for each build that we are running the same number of tests both before and after. This change could easy miss testing things and we would not easily know.

**Comment by saurav2608** [comment]
@jreback : I compared this build to one just before. Details - 

PR|platform | No of Tests | Link 
--|---------|------|--------|
|#24632    |Azure | 336,141|https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=8496&view=ms.vss-test-web.build-test-results-tab
|#25209 |Azure |336,141|https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=8495&view=ms.vss-test-web.build-test-results-tab

For Travis the order was a bit different. But I compared the this build with the one just before. And compared 4 builds

PR>build>Job|platform | No of Tests | Link 
--------------|---------|-------------|--------|
#24632 > Build #52470 > Job #52470.1 | Travis | 43095 passed, 2814 skipped, 1008 xfailed, 20 xpassed, 58 warnings in 678.23 seconds | https://travis-ci.org/pandas-dev/pandas/jobs/494931336
#25362 > Build #52468 > Job #52468.1 | Travis | 43095 passed, 2814 skipped, 1008 xfailed, 20 xpassed, 58 warnings in 659.78 seconds | https://travis-ci.org/pandas-dev/pandas/jobs/494917125
#24632 > Build #52470 > Job #52470.2 | Travis | 45407 passed, 1325 skipped, 783 xfailed, 20 xpassed, 24 warnings in 1144.70 seconds  | https://travis-ci.org/pandas-dev/pandas/jobs/494931338
#25362 > Build #52468 > Job #52468.2 | Travis | 45407 passed, 1325 skipped, 783 xfailed, 20 xpassed, 24 warnings in 1093.64 seconds  | https://travis-ci.org/pandas-dev/pandas/jobs/494917126
#24632 > Build #52470 > Job #52470.3 | Travis | 45338 passed, 917 skipped, 1013 xfailed, 20 xpassed, 42 warnings in 880.59 seconds| https://travis-ci.org/pandas-dev/pandas/jobs/494931340
#25362 > Build #52468 > Job #52468.3 | Travis | 45338 passed, 917 skipped, 1013 xfailed, 20 xpassed, 42 warnings in 833.46 seconds |https://travis-ci.org/pandas-dev/pandas/jobs/494917127
#24632 > Build #52470 > Job #52470.4 | Travis | 45410 passed, 842 skipped, 1014 xfailed, 21 xpassed, 90 warnings in 1305.06 seconds | https://travis-ci.org/pandas-dev/pandas/jobs/494931341
#25362 > Build #52468 > Job #52468.4 | Travis | 45410 passed, 842 skipped, 1014 xfailed, 21 xpassed, 90 warnings in 1243.80 seconds   |https://travis-ci.org/pandas-dev/pandas/jobs/494917128

File: ci/setup_env.sh
Comment (h-vetinari) [inline ci/setup_env.sh]: @saurav2608 
As long as `conda list` is called sometime after `activate pandas-dev`, everything is fine for what I'm trying to do in #25377. Happy to close my PR if this will move forward.

**Review by h-vetinari — COMMENTED** [review]
Didn't see this PR before, I'll close #25377 once this gets merged.

**Comment by jreback** [comment]
@saurav2608 ok this looks fine. I still am going to wait till we do 0.24.2, we will have to backport this as well I think. We have had CI brokeness in the past from even minor changes in the CI code.

**Comment by WillAyd** [comment]
Can you merge master on this?

**Comment by h-vetinari** [comment]
What's the status of this PR? I constantly need to look up manually which version is in failing posix environments. If this is still taking longer, can we merge #25377 in the interim?

**Comment by h-vetinari** [comment]
> What's the status of this PR? I constantly need to look up manually which version is in failing posix environments. If this is still taking longer, can we merge #25377 in the interim?

**Comment by jreback** [comment]
this needs a rebase and the PY2 and numpy bump merged first.

**Comment by jreback** [comment]
@saurav2608 if you wan to rebase this, i think we can do this now.

**Comment by saurav2608** [comment]
> @saurav2608 if you wan to rebase this, i think we can do this now.

@jreback - green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jreback and thanks @datapythonista for guiding this along.

**Pull request merged**
